pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/GroupByOptionsIntegrationTest.java

```diff
@@ -0,0 +1,681 @@
+
+    assertNumGroupsLimitException(query);
+  }
+
+  @Test
+  public void testExceptionIsThrownWhenErrorOnNumGroupsLimitHintIsSetAndLimitIsReachedV2()
+      throws Exception {
+    String query = " set numGroupsLimit=1;"
+        + " select /*+  aggOptions(error_on_num_groups_limit='true') */"
+        + " i, j, count(*) as cnt "
+        + " from " + getTableName()
+        + " group by i, j "
+        + " order by i, j ";
+
+    assertNumGroupsLimitException(query);
+  }
+
+  @Test
+  public void testExceptionIsThrownWhenErrorOnNumGroupsLimitOptionIsSetAndLimitIsReachedV1()
+      throws Exception {
+    String query = " set errorOnNumGroupsLimit=true; set numGroupsLimit=1;"
+        + " select i, j, count(*) as cnt "
+        + " from " + getTableName()
+        + " group by i, j "
+        + " order by i, j ";
+
+    assertNumGroupsLimitException(query);
+  }
+
+  @Test
+  public void testExceptionIsThrownWhenErrorOnNumGroupsLimitOptionIsSetAndLimitIsReachedV2()
+      throws Exception {
+    String query = " set errorOnNumGroupsLimit=true; "
+        + "select /*+  aggOptions(num_groups_limit='1') */ i, j, count(*) as cnt "
+        + " from " + getTableName()
+        + " group by i, j "
+        + " order by i, j ";
+
+    assertNumGroupsLimitException(query);
+  }
+
+  private void assertNumGroupsLimitException(String query)
+      throws Exception {
+    JsonNode result = postV2Query(query);
+
+    CharSeq errorMessage = toResultStr(result);
+
+    Assert.assertTrue(errorMessage.startsWith("QueryExecutionError:\n"
+            + "Received error query execution result block: {1000=NUM_GROUPS_LIMIT has been reached at "),
+        errorMessage.toString());
+  }
+
+  // for debug only
+  protected Properties getPinotConnectionProperties() {
+    Properties properties = new Properties();
+    properties.put("timeoutMs", "3600000");
+    properties.put("brokerReadTimeoutMs", "3600000");
+    properties.put("brokerConnectTimeoutMs", "3600000");
+    properties.putAll(getExtraQueryProperties());
+    return properties;
+  }
+
+  private JsonNode postV2Query(String query)
+      throws Exception {
+    return postQuery(query, getBrokerQueryApiUrl(getBrokerBaseApiUrl(), true), null,
+        getExtraQueryProperties());
+  }
+
+  private static @NotNull CharSeq toResultStr(JsonNode mainNode) {
+    if (mainNode == null) {
+      return new CharSeq(new StringBuilder("null"));
+    }
+    JsonNode node = mainNode.get(RESULT_TABLE);
+    if (node == null) {
+      return toErrorString(mainNode.get("exceptions"));
+    }
+    return toString(node);
+  }
+
+  private static @NotNull CharSeq toExplainStr(JsonNode mainNode) {
+    if (mainNode == null) {
+      return new CharSeq(new StringBuilder("null"));
+    }
+    JsonNode node = mainNode.get(RESULT_TABLE);
+    if (node == null) {
+      return toErrorString(mainNode.get("exceptions"));
+    }
+    return toExplainString(node);
+  }
+
+  static class CharSeq implements CharSequence {
```

Comment:
What is the reason to have this class?

Reply:
Reverted to good ol' String.

Reply:
Just curious to know the reason to add it. Just performance or did it provide some extra feature?

Reply:
It was just to avoid too much string concat or StringBuilder.toString() calls caused by StringBuilder not implementing equals. ( we need a reasonable mutable string class :) )